### PR TITLE
Specify that hashes are made out of base64 encoding

### PR DESCRIPTION
**Pull request by Huntsekkern**
**tl;dr**: proposition: hashes should now be made on arrays of base64 encoded strings which look like `["Base64{Foo}", "Base64{Bar}"]` instead of escaping the `"` and `\` within Foo and Bar

---

Hey everyone! 

While performing tests between backend-1 and frontend-1, we stumbled upon an issue that could lead to serious troubles later (possibly next semester). Here's a quick description of the issue:

When hashing the _exact same string_, Go and JavaScript do not yield the same value for very specific cases: escaped characters. We introduced this "bug" in PR #133 when we decided to escape any `"` (quote character) and `\` (backslash character) into `\"` and `\\` respectively.

If either fe1 or be1 decides to adopt the other team's solution, we don't know how fe2 and be2 would react to these changes. Moreover, if be3 is added next semester, those might encounter the exact same problem. Thus, here's our solution to the problem:

Current implementation:
```java
public String hashStrings(Array[String] strs) {
	// transform each string input in base64
	for (String str : strs) { str = escapeString(str); }
	// creates the array in string form
	String arr = ... // result: arr = ["abc", "\"time\":2"]
	// hash the array in string form
	String hashed = hash(arr);

	// return the base64 representation of the hash
	return base64Encode(hashed);
}

// transforms any `"` into `\"` and any `\` into `\\`
public String escapeString(String str) { /* ... */ }
```

Proposed implementation:
```java
public String hashStrings(Array[String] strs) {
	// transform each string input in base64
	for (String str : strs) { str = base64Encode(str); }
	// creates the array in string form
	String arr = ... // result: arr = ["abc=", "bde="]
	// hash the array in string form
	String hashed = hash(arr);

	// return the base64 representation of the hash
	return base64Encode(hashed);
}
```

Note that only the `for-loop` is modified. This way, we will never have to escape characters since every string is encoded in base64.

:information_source: with our solution, **any** string used for hashing is converted in base64 (for-loop). Another solution could be only to encode a subset of those strings. For example, the creation time (`int`) will never contain any problematic characters. The same could be said for `R` and `M` (bytes used for roll-calls and meetings when hashing).

:information_source: frontend-1 is working in a JavaScript environment while backend-1 is working in Go.

### CODE REVIEW

**Review by wouewouel — APPROVED** [review]

**Comment by lhmerino** [comment]
Thank you for the detailed explanation. Is there a specific reason why you decided to escape the characters in the first place?

**Comment by Huntsekkern** [comment]
> Thank you for the detailed explanation. This looks good to me. Is there a specific reason why you decided to escape the characters in the first place?

The detailed explanation is fully to be credited to @RaulinN :)

Escaping the characters was decided in PR #133 as per @pierluca quote: (and he stands by this decision in our slack thread)
> The limitations of JSON (as a format without canonical representation) were discussed at length throughout this project.
In this particular case, we settled (on Bryan's recommendation) on something that 'looks' like JSON but is stricter: an array of UTF-8 encoded strings (or numbers), with some characters escaped (", \) and without any whitespace between syntactic elements. Essentially it doesn't require any external dependencies (as opposed to all of the above suggestions) and can be easily implemented 'by hand', by escaping a couple of characters and adding delimiters.

I guess the real advantage lies in "doesn't require any external dependencies"

**Review by GabrielFleischer — APPROVED** [review]

**Comment by SFulpius** [comment]
Closing because we choose not to change the specs
